File: python/pubsub.py
from concurrent.futures import TimeoutError
from google.cloud import pubsub_v1 # pip install google google-cloud google-cloud-pubsub
from googleapiclient.discovery import build # pip install google-api-python-client
from httplib2 import Http
import json
from oauth2client.service_account import ServiceAccountCredentials # pip install oauth2
import os
from time import sleep
from typing import Any, Dict, Text



# TODO start
os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = 'dotted-marking-327507-277d58834909.json'

project_id      = 'dotted-marking-327507' # https://console.cloud.google.com/apis/api/chat.googleapis.com/hangouts-chat?project=dotted-marking-327507
topic_id        = 'Testbot'
subscription_id = 'Testbot'

credentials = ServiceAccountCredentials.from_json_keyfile_name(os.environ['GOOGLE_APPLICATION_CREDENTIALS'], 'https://www.googleapis.com/auth/chat.bot')
chat = build('chat', 'v1', http = credentials.authorize(Http()))

flow_control = pubsub_v1.types.FlowControl(max_messages = 10)
timeout = 30
# TODO end



def itElse(obj, key: Text, alt: Any):
    if isinstance(obj, dict):
        return obj[key] if (key in obj) else alt
    return alt



publisher = pubsub_v1.PublisherClient()
topic_path = publisher.topic_path(project_id, topic_id)


with pubsub_v1.SubscriberClient() as subscriber:
    subscription_path = subscriber.subscription_path(project_id, subscription_id)

    print(f'Topic:        {topic_path}')
    print(f'Subscription: {subscription_path}\n')


    def respond(text = None, cards = None, space = None, thread = None) -> Dict:
        parent = space
        body = {}

        if text is not None and len(text) > 0:
            body['text'] = text

        if cards is not None and len(cards) > 0:
            body['cards'] = cards

        if thread is not None:
            body['thread'] = {
                'name': thread
            }

        response = chat.spaces().messages().create(
            parent = parent,
            body = body
        ).execute()

        return response

    def typeof(x):
        return type(x)


    def callback(msg: pubsub_v1.subscriber.message.Message) -> None:
        data = None

        try:
            data = json.loads(msg.data.decode('UTF-8'))
        except Exception as e:
            print('[ERROR]', e)
            return

        type = itElse(data, 'type', 'NONE')

        time = itElse(data, 'eventTime', 0)

        message = itElse(data, 'message', {})
        message_name            = itElse(message, 'name', '')
        message_text            = itElse(message, 'text', '')
        message_cards           = itElse(message, 'cards', [])
        message_attachments     = itElse(message, 'attachment', [])
        message_attachment      = itElse(message, 'attachment', [{}])[0]
        message_attachment_type = itElse(message_attachment, 'contentType', '')

        user = itElse(data, 'user', {})
        user_name    = itElse(user, 'name', '')
        user_display = itElse(user, 'displayName', 'User')
        user_avatar  = itElse(user, 'avatarUrl', '')
        user_email   = itElse(user, 'email', '')
        user_type    = itElse(user, 'type', '')

        space = itElse(data, 'space', None)
        space_name   = itElse(space, 'name', '')
        space_type   = itElse(space, 'type', '')
        space_single = itElse(space, 'singleUserBotDm', False)

        thread = itElse(message, 'thread', {})
        thread_name = itElse(thread, 'name', None)

        print(f"Message:\n{data}")
        #print(f"Text:    {message_text}")
        #print(f"Space:   {space_name}")

        msg.ack()
        print()


        reply = ''
        cards = []

        if type == 'ADDED_TO_SPACE':
            if space_single:
                reply = f'Hey {user_display.split()[0]}, thanks for adding me!'

            else:
                reply = 'Hey everyone, thanks for adding me!'

        elif type == 'MESSAGE':
            if message_text.strip(' \n\t') == '@TestbotDC21':
                if not message_attachment_type:
                    if space_single:
                        reply = f'Yes?'
                    else:
                        reply = f'Yes, {user_display.split()[0]}?'

                message_text = ''

            if message_text:
                if space_single:
                    reply = f'_You said:_\n{message_text}'
                else:
                    reply = f'_<{user_name}> said:_\n{message_text}'
                    
            if message_text and message_attachment_type:
                reply += '\n\n'
            
            if message_attachment_type:
                if space_single:
                    reply += f'_You sent an attachment of type:_\n{message_attachment_type}'
                else:
                    reply += f'_<{user_name}> sent an attachment of type:_\n{message_attachment_type}'

        elif type == 'CARD_CLICKED':
            reply = ''

        elif type == 'REMOVED_FROM_SPACE':
            reply = ''

        elif type == 'CUSTOM':
            reply = '*Custom message recieved:*\n' + message_text
            cards = message_cards

            print(cards)
            print()

        if reply.strip(' \n\t') != '':
            bot_response = respond(text = reply, cards = cards, space = space_name, thread = thread_name)
            print(f'Reply:\n{bot_response}')

            if space_name != 'spaces/7gV80IAAAAE':
                feedback = f'*Replied in _{space_name}_ to _{user_name}_ (_{user_display}_)*\n\n'

                if message_text:
                    feedback += message_text + '\n\n- - - - -\n\n'
                
                feedback += reply + '\n\n= = = = = = = =\n\n'

                respond(text = feedback, cards = cards, space = 'spaces/7gV80IAAAAE', thread = None)

        else:
            print(f'No reply from bot after {type}.')


        print('\n' + '='*80 + '\n')
        


    while True:
        pull_future = subscriber.subscribe(
            subscription = subscription_path,
            callback     = callback,
            flow_control = flow_control,
        )

        try:
            print('Listening...\n' + '='*80 + '\n')
            pull_future.result(timeout = None)
        except TimeoutError:
            print(f'Error: TimeoutError')
            pull_future.cancel()  # Trigger the shutdown
            pull_future.result()  # Block until the shutdown is complete
        except Exception as e:
            print(f'Error: {e}')
            print()
            sleep(2)
            

File: nodejs/pubsub.js
require('dotenv').config()
const {google} = require('googleapis')
const {PubSub} = require('@google-cloud/pubsub');
const request = require('request');



// TODO start
const GOOGLE_APPLICATION_CREDENTIALS = process.env.GOOGLE_APPLICATION_CREDENTIALS;

project_id      = 'dotted-marking-327507'
topic_id        = 'Testbot'
subscription_id = 'Testbot'

const credentials = new google.auth.GoogleAuth({
    keyFile: GOOGLE_APPLICATION_CREDENTIALS,
    scopes: ['https://www.googleapis.com/auth/chat.bot'],
})
// TODO end



function itElse(obj, key, alt = null) {
    if (obj instanceof Array) {
        return (obj.length >= key + 1) ? obj[key] : alt
    } else if (obj instanceof Object) {
        return (obj.hasOwnProperty(key)) ? obj[key] : alt
    } else {
        return alt
    }
}


function isKey(obj, key) {
    if (obj instanceof Array) {
        return (obj.length >= key + 1)
    } else if (obj instanceof Object) {
        return obj.hasOwnProperty(key)
    } else {
        return false
    }
}



const pubSubClient = new PubSub();

function listenForMessages() {
    const subscription = pubSubClient.subscription(subscription_id)


    function respond(reply, verbose = true) {
        const data = {
            headers: {'content-type': 'application/json'},
            url:     'http://localhost:7007',
            body:    JSON.stringify(reply)
        }
        
        let response = ''

        request.post(data, function(error, response, body) {
            if (error) {
                console.error(`Error:\n${error}`)
            } else if (verbose && body) {
                console.log(`Reply:\n${body}`)
                console.log('\n' + '='.repeat(80) + '\n')
            }
        })

        return response
    }
  

    const callback = (msg) => {
        const data = JSON.parse(msg.data)

        const type = itElse(data, 'type', 'NONE').toUpperCase()

        const time = itElse(data, 'eventTime', 0)

        const message = itElse(data, 'message', {})
        const message_name            = itElse(message, 'name', '')
        const message_text            = itElse(message, 'text', '')
        const message_cards           = itElse(message, 'cards', [])
        const message_attachments     = itElse(message, 'attachment', [])
        const message_attachment      = itElse(message, 'attachment', [{}])[0]
        const message_attachment_type = itElse(message_attachment, 'contentType', '')

        const user = itElse(data, 'user', {})
        const user_name    = itElse(user, 'name', '')
        const user_display = itElse(user, 'displayName', 'User')
        const user_avatar  = itElse(user, 'avatarUrl', '')
        const user_email   = itElse(user, 'email', '')
        const user_type    = itElse(user, 'type', '')

        const space = itElse(data, 'space', null)
        const space_name   = itElse(space, 'name', '')
        const space_type   = itElse(space, 'type', '')
        const space_single = itElse(space, 'singleUserBotDm', false)

        const thread = itElse(message, 'thread', {})
        const thread_name = itElse(thread, 'name', null)

        console.log(`Message:\n${JSON.stringify(data)}`)

        msg.ack()
        console.log()


        reply = {}
        reply_text  = ''
        reply_image = ''
        reply_cards = []


        if (type == 'ADDED_TO_SPACE') {
            if (space_single) {
                reply = `Hey ${user_display.split()[0]}, thanks for adding me!`
            } else {
                reply = 'Hey everyone, thanks for adding me!'
            }
        } else if (type == 'MESSAGE') {
            reply_text = `_You said:_\n${message_text}`
        } else if (type == 'CARD_CLICKED') {
            reply = 'CARD_CLICKED'
        } else if (type == 'REMOVED_FROM_SPACE') {
            reply = 'This should not appear'
        } else {
            reply = type
        }

        if (reply_text) {
            reply['text'] = reply_text
        } if (reply_image) {
            reply['image'] = reply_image
        } if (reply_cards) {
            reply['cards'] = reply_cards
        }
        
        if (reply && space_name) {
            reply['recipient'] = space_name.split('/')[1]
        } else {
            return
        }
        

        const bot_response = respond(reply)

        if (space_name != 'spaces/7gV80IAAAAE') {
            let feedback = {}
            feedback['recipient'] = '7gV80IAAAAE'

            feedback['text'] = `*Replied in _${space_name}_ to _${user_name}_ (_${user_display}_)*\n\n`

            if (message_text) {
                feedback['text'] += message_text + '\n\n- - - - -\n\n'
            }
            
            feedback['text'] += reply_text + '\n\n= = = = = = = =\n\n'

            respond(feedback, false)
        }

    }
    
    subscription.on('message', callback)
}
  
listenForMessages()